modules/kinesis/src/it/scala/com/snowplowanalytics/snowplow/enrich/kinesis/EnrichKinesisSpec.scala

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright (c) 2022-2022 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.snowplow.enrich.kinesis
+
+import scala.concurrent.duration._
+
+import cats.effect.{Blocker, IO}
+import cats.effect.concurrent.Ref
+
+import fs2.Stream
+
+import retry.syntax.all._
+import retry.RetryPolicies
+
+import cats.effect.testing.specs2.CatsIO
+
+import org.specs2.mutable.Specification
+import org.specs2.specification.AfterAll
+
+import com.snowplowanalytics.snowplow.analytics.scalasdk.Event
+
+import com.snowplowanalytics.snowplow.enrich.common.fs2.config.io.{Input, Output}
+
+class EnrichKinesisSpec extends Specification with AfterAll with CatsIO {
+
+  override protected val Timeout = 10.minutes
+
+  val retryPolicy =
+    RetryPolicies.limitRetriesByCumulativeDelay(5.minutes, RetryPolicies.capDelay[IO](2.second, RetryPolicies.fullJitter[IO](1.second)))
+
+  def afterAll: Unit = Containers.localstack.stop()
+
+  "enrich-kinesis" should {
+    "initialize with the minimal config" in {
+      for {
+        enrich <- IO(Containers.enrichMinimal)
+        isRunning = IO(enrich.isRunning())
+        _ <- isRunning.retryingOnFailures(
+          _ == false,
+          retryPolicy,
+          (_, _) => IO(())
+        )
+      } yield {
+        enrich.getLogs() must contain("Running Enrich")
+      }
+    }
+
+    "emit the correct number of enriched events and bad rows" in {
+      type AggregateGood = List[Event]
+      type AggregateBad = List[String]
+      case class Aggregates(good: AggregateGood, bad: AggregateBad)
+
+      /** This function is in charge of:
+        * - Generating and sending collector payloads to the input stream of Enrich
+        * - Reading the enriched stream and bad stream populated by Enrich
+        * - Aggregating the above
+        */
+      def run(
```

Comment:
What do you think about slightly different version, I was thinking about sth like:

```scala
sealed trait OutputRow
object OutputRow {

  final case class Good(event: Event) extends OutputRow
  final case class Bad(line: String) extends OutputRow
}

final case class TestResources(rawSink: ByteSink[IO],
                               goodSource: Stream[IO, OutputRow.Good],
                               badSource: Stream[IO, OutputRow.Bad]

"TEST" in {
        createResources().use { resources =>
          val writingInput = generateInput(resources.rawSink, nbGood = 100, nbBad = 10)
          
          val readingOutput = 
            resources.goodSource
              .merge(resources.badSource)
              .interruptAfter(2.minutes)

          val output = readingOutput.concurrently(writingInput)

          val result = output.compile.toList
          
          //assert result list
        }
} 

  
  private def createResources(): Resource[IO, TestResources] = {
    for {
      blocker <- Blocker[IO]
      rawSink <- Sink.init[IO](blocker, rawConfig) 
    } yield {
      val goodSource = asGood(sourceStream(blocker, enrichedConfiged))
      val badSource = asBad(sourceStream(blocker, badConfig))
      TestResources(rawSink, goodSource, badSource)
    }
  }
  
  private def generateInput(sink: ByteSink[IO], nbGood: Long, nbBad: Long): Stream[IO, Unit] = {
    CollectorPayloadGen.generate[IO](nbGood, nbBad)
      .evalMap(events => sink(List(events)))
  }

  private def sourceStream(blocker: Blocker, config: Input.Kinesis): Stream[IO, Array[Byte]] =
    Source.init[IO](blocker, config, ConfigCount.monitoring)
      .map(KinesisRun.getPayload)
      
  private def asGood(source: Stream[IO, Array[Byte]]): Stream[IO, OutputRow.Good] = {
    source.map { bytes => 
      OutputRow.Good {
        Event.parse(new String(bytes))
          .getOrElse(throw new RuntimeException("can't parse enriched event"))
      }
    }
  }
  private def asBad(source: Stream[IO, Array[Byte]]): Stream[IO, OutputRow.Bad] =
   source.map(bytes => OutputRow.Bad(new String(bytes)))
```

I usually try to make test content as short as possible, but with clear input and output. All test helpers extracted somewhere else, checking impl details only when needed.

Reply:
Yeah I also don't like this very long test, I'll update